torchtitan/parallelisms/parallelize_llama.py

```diff
@@ -324,17 +324,23 @@ def apply_fsdp(
         )
     fully_shard(model, **fsdp_config, reshard_after_forward=not pp_enabled)
 
-    if pp_enabled:
-        # TODO
-        # This PR https://github.com/pytorch/pytorch/pull/129519 added a safety check to avoid using 2D/3D DCP since
-        # without strided sharding, DCP can not safely support resharding for 2D/3D.  However, for PP to work, even
-        # without resharding, we load a seed-checkpoint and need to disable the safety mechanism.  This hack should be
-        # removed after strided sharding is landed in DCP.
-        for module in model.modules():
-            assert len(module._load_state_dict_pre_hooks) <= 1
+    # TODO: This PR https://github.com/pytorch/pytorch/pull/129519 added a safety check
+    # to avoid using 2D/3D DCP since without strided sharding, DCP can not safely support
+    # resharding for 2D/3D. We keep this safety check disablement here for now and will
+    # remove it when PyTorch gets the next minor release (PyTorch 2.5).
+    for module in model.modules():
+        assert len(module._load_state_dict_pre_hooks) <= 1
+        if len(module._load_state_dict_pre_hooks) == 1:
+            logger.warning(
+                "a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a "
+                "version newer than 2024-08-09 nightly release to include the change in "
+                "https://github.com/pytorch/pytorch/pull/130760"
+            )
             module._load_state_dict_pre_hooks.clear()
-            assert len(module._state_dict_pre_hooks) <= 1
-            module._state_dict_pre_hooks.clear()
+
+        assert len(module._state_dict_pre_hooks) <= 1
+        module._state_dict_pre_hooks.clear()
+
```

Comment:
I added a check here to print warning if some hooks are detected. I expected to see no hooks so no warning would be printed. However, I saw in the console that this warning actually appeared:
```
[rank0]:2024-08-09 11:53:01,498 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank0]:2024-08-09 11:53:01,498 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank1]:2024-08-09 11:53:01,492 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank1]:2024-08-09 11:53:01,492 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank3]:2024-08-09 11:53:02,163 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank3]:2024-08-09 11:53:02,164 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank6]:2024-08-09 11:53:02,155 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank6]:2024-08-09 11:53:02,155 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank2]:2024-08-09 11:53:02,164 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank2]:2024-08-09 11:53:02,164 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank4]:2024-08-09 11:53:02,110 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank4]:2024-08-09 11:53:02,110 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank7]:2024-08-09 11:53:02,151 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank7]:2024-08-09 11:53:02,151 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank5]:2024-08-09 11:53:02,105 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
[rank5]:2024-08-09 11:53:02,105 - root - WARNING - a safety check on 2D/3D DCP is detected. Please upgrade PyTorch to a version newer than 2024-08-09 nightly release to include the change in https://github.com/pytorch/pytorch/pull/130760
```

Since FSDP no longer adds the `_raise_not_implemented_if_2d` [hook](https://github.com/pytorch/pytorch/pull/129519/files#diff-8df0199828008c13ead0271f0f6a6bad55fd8c3944ca915cd12a157f8675af93R157), this check is not supposed to be `True`. Or, was the PR (https://github.com/pytorch/torchtitan/pull/460) removing some hooks that are unrelated to 2D/3D disablement?

cc @fegin @wconstab

Reply:
Good question, I don't know. Can you print out the hook to debug? Since this is Python, it should give you some reasonable hint about what is this hook.